internal/cloud/azure/cloud.go

```diff
@@ -0,0 +1,428 @@
+// metadata interface.
+func (c *Cloud) GetLoadBalancerEndpoint(ctx context.Context) (string, error) {
+	resourceGroup, err := c.imds.resourceGroup(ctx)
+	if err != nil {
+		return "", fmt.Errorf("retrieving resource group: %w", err)
+	}
+	uid, err := c.imds.uid(ctx)
+	if err != nil {
+		return "", fmt.Errorf("retrieving instance UID: %w", err)
+	}
+
+	lb, err := c.getLoadBalancer(ctx, resourceGroup, uid)
+	if err != nil {
+		return "", fmt.Errorf("retrieving load balancer: %w", err)
+	}
+	if lb == nil || lb.Properties == nil {
+		return "", errors.New("could not dereference load balancer IP configuration")
+	}
+
+	var pubIP string
+	for _, fipConf := range lb.Properties.FrontendIPConfigurations {
+		if fipConf == nil || fipConf.Properties == nil || fipConf.Properties.PublicIPAddress == nil || fipConf.Properties.PublicIPAddress.ID == nil {
+			continue
+		}
+		pubIP = path.Base(*fipConf.Properties.PublicIPAddress.ID)
+		break
+	}
+
+	resp, err := c.pubIPAPI.Get(ctx, resourceGroup, pubIP, nil)
+	if err != nil {
+		return "", fmt.Errorf("retrieving load balancer public IP address: %w", err)
+	}
+	if resp.Properties == nil || resp.Properties.IPAddress == nil {
+		return "", fmt.Errorf("could not resolve public IP address reference for load balancer")
+	}
+	return *resp.Properties.IPAddress, nil
+}
+
+// List retrieves all instances belonging to the current constellation.
+func (c *Cloud) List(ctx context.Context) ([]metadata.InstanceMetadata, error) {
+	resourceGroup, err := c.imds.resourceGroup(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving resource group: %w", err)
+	}
+
+	uid, err := c.imds.uid(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving instance UID: %w", err)
+	}
+
+	instances := []metadata.InstanceMetadata{}
+	pager := c.scaleSetsAPI.NewListPager(resourceGroup, nil)
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("retrieving scale sets: %w", err)
+		}
+
+		for _, scaleSet := range page.Value {
+			if scaleSet == nil || scaleSet.Name == nil || scaleSet.Tags == nil ||
+				scaleSet.Tags[cloud.TagUID] == nil || *scaleSet.Tags[cloud.TagUID] != uid {
+				continue
+			}
+			vmPager := c.scaleSetsVMAPI.NewListPager(resourceGroup, *scaleSet.Name, nil)
+			for vmPager.More() {
+				vmPage, err := vmPager.NextPage(ctx)
+				if err != nil {
+					return nil, fmt.Errorf("retrieving vms: %w", err)
+				}
+
+				for _, vm := range vmPage.Value {
+					if vm == nil || vm.InstanceID == nil {
+						continue
+					}
+					interfaces, err := c.getVMInterfaces(ctx, *vm, resourceGroup, *scaleSet.Name, *vm.InstanceID)
+					if err != nil {
+						return nil, fmt.Errorf("retrieving VM network interfaces: %w", err)
+					}
+					instance, err := convertToInstanceMetadata(*vm, interfaces)
+					if err != nil {
+						return nil, fmt.Errorf("converting VM to instance metadata: %w", err)
+					}
+					instances = append(instances, instance)
+				}
+			}
```

Comment:
nit: maybe we could first collect scale sets and list instances afterwards to reduce nesting?

Reply:
While I agree this would reduce nesting, this would also negatively impact performance since we would iterate the list twice, once to collect and once to process, which is unneeded imo.